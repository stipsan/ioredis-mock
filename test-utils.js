// Runs both a regular string suite and a buffer suite
export function runTwinSuite(commandName, cb) {
  const commands = [commandName, `${commandName}Buffer`]

  return commands.forEach(command =>
    command === commandName
      ? cb(command, (a, b) => a === b)
      : cb(command, (a, b) => a.equals(Buffer.from(b)))
  )
}
