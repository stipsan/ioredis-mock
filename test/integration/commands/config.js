import Redis from 'ioredis'

// eslint-disable-next-line import/no-relative-parent-imports
import { runTwinSuite } from '../../../test-utils'

runTwinSuite('config', (command, equals) => {
  ;(process.env.IS_BROWSER && command.endsWith('Buffer')
    ? describe.skip
    : describe)(command, () => {
    const redis = new Redis()

    afterAll(() => {
      redis.disconnect()
    })

    it('should throw on too few arguments', async () => {
      expect.hasAssertions()

      try {
        await redis[command]()
      } catch (err) {
        expect(err.message).toMatchSnapshot()
      }
    })

    it('should throw on unknown subcommand', async () => {
      expect.hasAssertions()

      try {
        await redis[command]('foobar')
      } catch (err) {
        expect(err.message).toMatchSnapshot()
      }
    })

    describe('get', () => {
      it('should throw on wrong number of arguments', async () => {
        expect.hasAssertions()

        try {
          await redis[command]('get')
        } catch (err) {
          expect(err.message).toMatchSnapshot()
        }
      })

      it('returns a list over config options', async () => {
        expect(await redis[command]('GET', '*')).toEqual(expect.any(Array))
      })
    })

    describe('set', () => {
      it('should throw on wrong number of arguments', async () => {
        expect.assertions(2)

        try {
          await redis[command]('set')
        } catch (err) {
          expect(err.message).toMatchSnapshot()
        }

        try {
          await redis[command]('SET', 'foo')
        } catch (err) {
          expect(err.message).toMatchSnapshot()
        }
      })
      ;(process.env.IS_E2E ? it.skip : it)(
        'should throw as we actually do not support setting the config',
        async () => {
          expect.hasAssertions()

          try {
            await redis[command]('SET', 'maxmemory', '1000000')
          } catch (err) {
            expect(err.message).toMatchSnapshot()
          }
        }
      )
      it('should throw as the option does not exist', async () => {
        expect.hasAssertions()

        try {
          await redis[command]('SET', 'maxmockmemory', '1000000')
        } catch (err) {
          expect(err.message).toMatchSnapshot()
        }
      })
    })

    describe('resetstat', () => {
      it('should throw on wrong number of arguments', async () => {
        expect.hasAssertions()

        try {
          await redis[command]('RESETSTAT', 'foo')
        } catch (err) {
          expect(err.message).toMatchSnapshot()
        }
      })

      it('returns OK', async () => {
        expect(equals(await redis[command]('RESETSTAT'), 'OK')).toBe(true)
      })
    })

    describe('rewrite', () => {
      it('should throw on wrong number of arguments', async () => {
        expect.hasAssertions()

        try {
          await redis[command]('REWRITE', 'foo')
        } catch (err) {
          expect(err.message).toMatchSnapshot()
        }
      })

      it('throws as redis is running without a config file', async () => {
        expect.hasAssertions()

        try {
          await redis[command]('REWRITE')
        } catch (err) {
          expect(err.message).toMatchSnapshot()
        }
      })
    })

    describe('help', () => {
      it('should throw on wrong number of arguments', async () => {
        expect.assertions(2)

        try {
          await redis[command]('HELP', 'foo')
        } catch (err) {
          expect(err.message).toMatchSnapshot()
        }

        try {
          await redis[command]('HELP', 'foo', 'bar')
        } catch (err) {
          expect(err.message).toMatchSnapshot()
        }
      })

      it('prints a list over available subcommands', async () => {
        const result = await redis[command]('HELP')

        expect(result).toMatchSnapshot()
      })
    })
  })
})
