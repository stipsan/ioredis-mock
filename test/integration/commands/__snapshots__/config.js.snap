// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`config get should throw on wrong number of arguments 1`] = `"ERR wrong number of arguments for 'config|get' command"`;

exports[`config help prints a list over available subcommands 1`] = `
Array [
  "CONFIG <subcommand> [<arg> [value] [opt] ...]. Subcommands are:",
  "GET <pattern>",
  "    Return parameters matching the glob-like <pattern> and their values.",
  "SET <directive> <value>",
  "    Set the configuration <directive> to <value>.",
  "RESETSTAT",
  "    Reset statistics reported by the INFO command.",
  "REWRITE",
  "    Rewrite the configuration file.",
  "HELP",
  "    Prints this help.",
]
`;

exports[`config help should throw on wrong number of arguments 1`] = `"ERR wrong number of arguments for 'config|help' command"`;

exports[`config help should throw on wrong number of arguments 2`] = `"ERR wrong number of arguments for 'config|help' command"`;

exports[`config resetstat should throw on wrong number of arguments 1`] = `"ERR wrong number of arguments for 'config|resetstat' command"`;

exports[`config rewrite should throw on wrong number of arguments 1`] = `"ERR wrong number of arguments for 'config|rewrite' command"`;

exports[`config rewrite throws as redis is running without a config file 1`] = `"ERR The server is running without a config file"`;

exports[`config set should throw as the option does not exist 1`] = `"ERR Unknown option or number of arguments for CONFIG SET - 'maxmockmemory'"`;

exports[`config set should throw as we actually do not support setting the config 1`] = `"ERR Unknown option or number of arguments for CONFIG SET - 'maxmemory'"`;

exports[`config set should throw on wrong number of arguments 1`] = `"ERR wrong number of arguments for 'config|set' command"`;

exports[`config set should throw on wrong number of arguments 2`] = `"ERR wrong number of arguments for 'config|set' command"`;

exports[`config should throw on too few arguments 1`] = `"ERR wrong number of arguments for 'config' command"`;

exports[`config should throw on unknown subcommand 1`] = `"ERR unknown subcommand 'foobar'. Try CONFIG HELP."`;

exports[`configBuffer get should throw on wrong number of arguments 1`] = `"ERR wrong number of arguments for 'config|get' command"`;

exports[`configBuffer help prints a list over available subcommands 1`] = `
Array [
  Buffer<
  43 4f 4e 46 49 47 20 3c 73 75 62 63 6f 6d 6d 61 6e 64 3e 20 5b 3c 61 72 67 3e
  20 5b 76 61 6c 75 65 5d 20 5b 6f 70 74 5d 20 2e 2e 2e 5d 2e 20 53 75 62 63 6f
  6d 6d 61 6e 64 73 20 61 72 65 3a
>,
  Buffer<47 45 54 20 3c 70 61 74 74 65 72 6e 3e>,
  Buffer<
  20 20 20 20 52 65 74 75 72 6e 20 70 61 72 61 6d 65 74 65 72 73 20 6d 61 74 63
  68 69 6e 67 20 74 68 65 20 67 6c 6f 62 2d 6c 69 6b 65 20 3c 70 61 74 74 65 72
  6e 3e 20 61 6e 64 20 74 68 65 69 72 20 76 61 6c 75 65 73 2e
>,
  Buffer<
  53 45 54 20 3c 64 69 72 65 63 74 69 76 65 3e 20 3c 76 61 6c 75 65 3e
>,
  Buffer<
  20 20 20 20 53 65 74 20 74 68 65 20 63 6f 6e 66 69 67 75 72 61 74 69 6f 6e 20
  3c 64 69 72 65 63 74 69 76 65 3e 20 74 6f 20 3c 76 61 6c 75 65 3e 2e
>,
  Buffer<52 45 53 45 54 53 54 41 54>,
  Buffer<
  20 20 20 20 52 65 73 65 74 20 73 74 61 74 69 73 74 69 63 73 20 72 65 70 6f 72
  74 65 64 20 62 79 20 74 68 65 20 49 4e 46 4f 20 63 6f 6d 6d 61 6e 64 2e
>,
  Buffer<52 45 57 52 49 54 45>,
  Buffer<
  20 20 20 20 52 65 77 72 69 74 65 20 74 68 65 20 63 6f 6e 66 69 67 75 72 61 74
  69 6f 6e 20 66 69 6c 65 2e
>,
  Buffer<48 45 4c 50>,
  Buffer<20 20 20 20 50 72 69 6e 74 73 20 74 68 69 73 20 68 65 6c 70 2e>,
]
`;

exports[`configBuffer help should throw on wrong number of arguments 1`] = `"ERR wrong number of arguments for 'config|help' command"`;

exports[`configBuffer help should throw on wrong number of arguments 2`] = `"ERR wrong number of arguments for 'config|help' command"`;

exports[`configBuffer resetstat should throw on wrong number of arguments 1`] = `"ERR wrong number of arguments for 'config|resetstat' command"`;

exports[`configBuffer rewrite should throw on wrong number of arguments 1`] = `"ERR wrong number of arguments for 'config|rewrite' command"`;

exports[`configBuffer rewrite throws as redis is running without a config file 1`] = `"ERR The server is running without a config file"`;

exports[`configBuffer set should throw as the option does not exist 1`] = `"ERR Unknown option or number of arguments for CONFIG SET - 'maxmockmemory'"`;

exports[`configBuffer set should throw as we actually do not support setting the config 1`] = `"ERR Unknown option or number of arguments for CONFIG SET - 'maxmemory'"`;

exports[`configBuffer set should throw on wrong number of arguments 1`] = `"ERR wrong number of arguments for 'config|set' command"`;

exports[`configBuffer set should throw on wrong number of arguments 2`] = `"ERR wrong number of arguments for 'config|set' command"`;

exports[`configBuffer should throw on too few arguments 1`] = `"ERR wrong number of arguments for 'config' command"`;

exports[`configBuffer should throw on unknown subcommand 1`] = `"ERR unknown subcommand 'foobar'. Try CONFIG HELP."`;
